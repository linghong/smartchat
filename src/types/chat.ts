export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
