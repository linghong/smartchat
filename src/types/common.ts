export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
