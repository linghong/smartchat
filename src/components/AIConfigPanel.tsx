import React, { ChangeEvent, FormEvent } from 'react';

import {
  Card,
  CardContent,
  CardHeader,
  CardTitle
} from '@/src/components/ui/card';
import { Input } from '@/src/components/ui/input';
import { Textarea } from '@/src/components/ui/textarea';
import { Button } from '@/src/components/ui/button';
import CustomSelect from '@/src/components/CustomSelect';
import { Slider } from '@/src/components/ui/slider';
import { Label } from '@/src/components/ui/label';

import { OptionType } from '@/src/types/common';
import { AIConfig } from '@/src/types/chat';

interface AIConfigPanelProps {
  selectedModel: OptionType;
  handleModelChange: (newValue: OptionType) => void;
  selectedNamespace: OptionType | null;
  handleNamespaceChange: (selectedOption: OptionType | null) => void;
  aiConfig: AIConfig;
  setAIConfig: (config: AIConfig) => void;
  modelOptions: OptionType[];
  fileCategoryOptions: OptionType[];
}

export default function AIConfigPanel({
  selectedModel,
  handleModelChange,
  selectedNamespace,
  handleNamespaceChange,
  aiConfig,
  setAIConfig,
  modelOptions,
  fileCategoryOptions
}: AIConfigPanelProps) {
  const handleInputChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const { id, value } = e.target;
    setAIConfig({ ...aiConfig, [id]: value });
  };

  const handleSliderChange = (id: string) => (value: number[]) => {
    setAIConfig({ ...aiConfig, [id]: value[0] });
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();

    console.log('Form submitted', {
      aiConfig,
      selectedNamespace
    });
  };

  return (
    <Card className="w-full mx-auto shadow-lg bg-white">
      <CardHeader className="bg-gradient-to-r from-indigo-50 via-gray-200  to-blue-100 border-b border-gray-300  shadow-sm shadow-cyan-200 drop-shadow-sm">
        <CardTitle className="text-center text-gray-800">
          Configure AI Assistant
        </CardTitle>
      </CardHeader>
      <CardContent className="bg-slate-100">
        <form onSubmit={handleSubmit} className="pt-6 pb-2 space-y-6">
          <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
            <div className="space-y-2">
              <Label htmlFor="name" className="text-gray-700">
                Name
              </Label>
              <Input
                id="name"
                value={aiConfig.name}
                onChange={handleInputChange}
                placeholder="Enter assistant name"
                className="w-full bg-gray-50"
                required
              />
            </div>

            <div className="space-y-2">
              <Label htmlFor="role">Assistant Role</Label>
              <Input
                id="role"
                value={aiConfig.role}
                onChange={handleInputChange}
                placeholder="Enter assistant role"
                className="w-full bg-gray-50"
                required
              />
            </div>
          </div>

          <div className="space-y-2">
            <Label htmlFor="namespace">Select RAG Namespace</Label>
            <CustomSelect
              id="namespace"
              options={[
                { value: 'none', label: 'None' },
                ...fileCategoryOptions
              ]}
              value={selectedNamespace?.value || 'none'}
              onChange={value => {
                const newSelectedNamespace = [
                  { value: 'none', label: 'None' },
                  ...fileCategoryOptions
                ].find(option => option.value === value);
                handleNamespaceChange(
                  value !== 'none' ? newSelectedNamespace || null : null
                );
              }}
              placeholder="Select RAG namespace"
            />
          </div>

          <div className="bg-slate-50 p-4 rounded-lg space-y-4">
            <div className="space-y-2">
              <label htmlFor="model" className="text-gray-700">
                Choose AI Model
              </label>
              <CustomSelect
                id="model"
                options={modelOptions}
                value={selectedModel.value}
                onChange={value => {
                  const newSelectedModel = modelOptions.find(
                    option => option.value === value
                  );
                  if (newSelectedModel) {
                    handleModelChange(newSelectedModel);
                  }
                }}
                placeholder="Select a model"
              />
            </div>

            <div className="space-y-2">
              <Label htmlFor="topP" className="text-gray-700">
                Model Top P: {aiConfig.topP.toFixed(1)}
              </Label>
              <Slider
                id="topP"
                min={0}
                max={1}
                step={0.1}
                value={[aiConfig.topP]}
                onValueChange={handleSliderChange('topP')}
                className="w-full"
              />
            </div>

            <div className="space-y-2">
              <Label htmlFor="temperature" className="text-gray-700">
                Model Temperature: {aiConfig.temperature.toFixed(1)}
              </Label>
              <Slider
                id="temperature"
                min={0}
                max={1}
                step={0.01}
                value={[aiConfig.temperature]}
                onValueChange={handleSliderChange('temperature')}
                className="w-full"
              />
            </div>
          </div>

          <div className="space-y-2">
            <Label htmlFor="basePrompt" className="text-gray-700">
              Enter Information You Want AI to Remember
            </Label>
            <Textarea
              id="basePrompt"
              rows={4}
              value={aiConfig.basePrompt}
              onChange={handleInputChange}
              placeholder="Enter text here for AI to remember throughout the chat"
              className="w-full resize-none bg-gray-50"
            />
          </div>

          <Button
            type="submit"
            className="w-full bg-gradient-to-r from-indigo-100 via-gray-800  to-slate-600 text-white font-semibold py-2 px-4 rounded"
          >
            Save Configuration
          </Button>
        </form>
      </CardContent>
    </Card>
  );
}
