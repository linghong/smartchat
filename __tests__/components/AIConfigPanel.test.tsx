import React, { act } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import AIConfigPanel from '@/src/components/AIConfigPanel';
import { OptionType } from '@/src/types/common';

// Mock Next.js router
jest.mock('next/router', () => ({
  useRouter: () => ({
    push: jest.fn()
  })
}));

// Mock the postAIConfig function
jest.mock('@/src/utils/sqliteAIConfigApiClient', () => ({
  postAIConfig: jest.fn(),
  getAIConfigs: jest.fn().mockResolvedValue([])
}));

// Mock the UI components
jest.mock('@/src/components/ui/card', () => ({
  Card: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  CardHeader: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  CardTitle: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
  CardContent: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  )
}));

jest.mock('@/src/components/CustomSelect', () => ({
  __esModule: true,
  default: ({
    id,
    options,
    value,
    onChange
  }: {
    id: string;
    options: OptionType[];
    value: string;
    onChange: (value: string) => void;
  }) => (
    <select
      id={id}
      value={value}
      onChange={e => onChange(e.target.value)}
      aria-label={id}
    >
      {options.map(option => (
        <option key={option.value} value={option.value}>
          {option.label}
        </option>
      ))}
    </select>
  )
}));

jest.mock('@/src/components/ui/slider', () => ({
  Slider: () => <div>Slider</div>
}));

jest.mock('@/src/components/Notification', () => ({
  __esModule: true,
  default: ({ type, message }: { type: string; message: string }) => (
    <div data-testid={`notification-${type}`}>{message}</div>
  )
}));

// Mock the ChatContext
jest.mock('@/src/context/ChatContext', () => ({
  useChatContext: () => ({
    selectedModel: { value: 'gpt-4', label: 'GPT-4' }
  })
}));

describe('AIConfigPanel', () => {
  const mockAIConfig = {
    name: 'TestBot',
    role: 'Test Assistant',
    model: {
      value: 'gpt-4',
      label: 'GPT-4',
      category: 'openai',
      contextWindow: 128000,
      vision: true
    },
    basePrompt: 'Test prompt',
    topP: 0.7,
    temperature: 0.5
  };

  const mockProps = {
    handleModelChange: jest.fn(),
    selectedNamespace: { value: 'namespace1', label: 'Namespace 1' },
    handleNamespaceChange: jest.fn(),
    aiConfig: mockAIConfig,
    setAIConfig: jest.fn(),
    modelOptions: [
      { value: 'model1', label: 'Model 1' },
      { value: 'model2', label: 'Model 2' }
    ],
    fileCategoryOptions: [
      { value: 'namespace1', label: 'Namespace 1' },
      { value: 'namespace2', label: 'Namespace 2' }
    ]
  };

  beforeEach(() => {
    jest.clearAllMocks();
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => 'mock-token')
      },
      writable: true
    });
  });

  it('renders the component correctly', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    expect(screen.getByText('Configure AI Assistant')).toBeInTheDocument();
    expect(screen.getByLabelText('Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Assistant Role')).toBeInTheDocument();
    expect(screen.getByText('Choose AI Model')).toBeInTheDocument();
    expect(screen.getByText('Select RAG Namespace')).toBeInTheDocument();
    expect(
      screen.getByText('Enter Information You Want AI to Remember')
    ).toBeInTheDocument();
  });

  it('renders the Select components with correct options', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    expect(screen.getByText('Model 1')).toBeInTheDocument();
    expect(screen.getByText('Model 2')).toBeInTheDocument();
    expect(screen.getByText('None')).toBeInTheDocument();
    expect(screen.getByText('Namespace 1')).toBeInTheDocument();
    expect(screen.getByText('Namespace 2')).toBeInTheDocument();
  });

  it('renders the textarea with correct value', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const textarea = screen.getByPlaceholderText(
      'Enter text here for AI to remember throughout the chat'
    );
    expect(textarea).toHaveValue('Test prompt');
  });

  it('calls setAIConfig when input values change', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const nameInput = screen.getByPlaceholderText('Enter assistant name');
    await act(async () => {
      fireEvent.change(nameInput, { target: { value: 'New Name' } });
    });
    expect(mockProps.setAIConfig).toHaveBeenCalledWith(expect.any(Function));

    // Get the function passed to setAIConfig
    const setAIConfigArg = mockProps.setAIConfig.mock.calls[0][0];

    const mockPrevState = { ...mockProps.aiConfig };
    const result = setAIConfigArg(mockPrevState);

    expect(result).toEqual(expect.objectContaining({ name: 'New Name' }));
  });

  it('calls handleModelChange when model is selected', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const select = screen.getByLabelText('Choose AI Model');
    await act(async () => {
      fireEvent.change(select, { target: { value: 'model2' } });
    });
    expect(mockProps.handleModelChange).toHaveBeenCalledWith({
      value: 'model2',
      label: 'Model 2'
    });
  });

  it('calls handleNamespaceChange when namespace is selected', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });

    const select = screen.getByLabelText('Select RAG Namespace');
    await act(async () => {
      fireEvent.change(select, { target: { value: 'namespace2' } });
    });
    expect(mockProps.handleNamespaceChange).toHaveBeenCalledWith({
      value: 'namespace2',
      label: 'Namespace 2'
    });
  });

  it('submits the form with correct data', async () => {
    const { postAIConfig } = require('@/src/utils/sqliteAIConfigApiClient');
    postAIConfig.mockResolvedValue({ message: 'AI Config saved successfully' });

    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const submitButton = screen.getByText('Save Configuration');
    await act(async () => {
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(postAIConfig).toHaveBeenCalledWith(
        'mock-token',
        mockAIConfig,
        'namespace1'
      );
      expect(screen.getByTestId('notification-success')).toBeInTheDocument();
      expect(screen.getByTestId('notification-success')).toHaveTextContent(
        'AI Config saved successfully'
      );
    });
  });

  it('shows error notification on submission failure', async () => {
    const { postAIConfig } = require('@/src/utils/sqliteAIConfigApiClient');
    postAIConfig.mockRejectedValue(new Error('Failed to save AI Config'));

    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const submitButton = screen.getByText('Save Configuration');
    await act(async () => {
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(screen.getByTestId('notification-error')).toBeInTheDocument();
      expect(screen.getByTestId('notification-error')).toHaveTextContent(
        'Failed to save AI Config'
      );
    });
  });

  it('redirects to login page if token is not present', async () => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => null)
      },
      writable: true
    });
    const mockPush = jest.fn();
    jest.spyOn(require('next/router'), 'useRouter').mockImplementation(() => ({
      push: mockPush
    }));
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const submitButton = screen.getByText('Save Configuration');
    await act(async () => {
      fireEvent.click(submitButton);
    });
    expect(mockPush).toHaveBeenCalledWith('/login');
  });

  it('disables submit button when required fields are empty', async () => {
    await act(async () => {
      const emptyConfig = { ...mockAIConfig, name: '', role: '' };
      render(<AIConfigPanel {...mockProps} aiConfig={emptyConfig} />);
    });
    const submitButton = screen.getByText('Save Configuration');
    expect(submitButton).toHaveAttribute('disabled');
  });

  it('shows error when submitting with existing name', async () => {
    const {
      getAIConfigs,
      postAIConfig
    } = require('@/src/utils/sqliteAIConfigApiClient');
    getAIConfigs.mockResolvedValue([{ name: 'TestBot' }]);
    postAIConfig.mockRejectedValue(
      new Error('An AI assistant with this name already exists.')
    );

    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });
    const submitButton = screen.getByText('Save Configuration');
    await act(async () => {
      fireEvent.click(submitButton);
    });
    await waitFor(() => {
      expect(screen.getByTestId('notification-error')).toBeInTheDocument();
      expect(screen.getByTestId('notification-error')).toHaveTextContent(
        'An AI assistant with this name already exists.'
      );
    });
  });

  it('updates topP and temperature sliders correctly', async () => {
    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });

    const topPSlider = screen.getByText('Model Top P: 0.70');
    const temperatureSlider = screen.getByText('Model Temperature: 0.50');

    expect(topPSlider).toBeInTheDocument();
    expect(temperatureSlider).toBeInTheDocument();
  });

  it('clears name field after successful submission', async () => {
    const { postAIConfig } = require('@/src/utils/sqliteAIConfigApiClient');
    postAIConfig.mockResolvedValue({ message: 'AI Config saved successfully' });

    await act(async () => {
      render(<AIConfigPanel {...mockProps} />);
    });

    const nameInput = screen.getByLabelText('Name');
    const submitButton = screen.getByText('Save Configuration');

    await act(async () => {
      fireEvent.click(submitButton);
    });
  });

  it('matches snapshot', async () => {
    let asFragment;
    await act(async () => {
      const rendered = render(<AIConfigPanel {...mockProps} />);
      asFragment = rendered.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
