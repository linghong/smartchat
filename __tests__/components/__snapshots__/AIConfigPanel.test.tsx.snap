// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIConfigPanel matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="
    transition-[max-height] duration-500 ease-in-out
    }
  "
  >
    <div
      class="flex flex-col w-full bg-gray-50 border-2 border-indigo-100 p-4 shadow-lg shadow-slate-200"
    >
      <div
        class="flex font-bold text-xl justify-center"
      >
        Config My AI Assistant
      </div>
      <div
        class="flex flex-col lg:flex-row w-full justify-between"
      >
        <div>
          <span>
            Choose AI Model:
          </span>
          <span>
            Model 1
          </span>
        </div>
        <div>
          <span>
            Select RAG File:
          </span>
          <span>
            Namespace 1
          </span>
        </div>
      </div>
      <div
        class="flex flex-col w-full items-center py-2"
      >
        <label
          class="text-base font-bold"
          for="userSystemPrompt"
        >
          Enter text here for AI to remember throughout the chat:
        </label>
        <textarea
          aria-label="Enter text here for AI to remember throughout the chat"
          class="w-full placeholder-gray-400 my-2 p-2 border-2 border-stone-300 rounded focus:ring-stone-100 focus:outline-none hover:bg-stone-50"
          id="userSystemPrompt"
          name="userSystemPrompt"
          rows="3"
        >
          Test prompt
        </textarea>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
